File: src/hello_world_cpp_pkg/src/custom_node.cpp
#include "rclcpp/rclcpp.hpp"

class CustomNode : public rclcpp::Node
{
private:
    int counter = 0;
    rclcpp::TimerBase::SharedPtr timer;

    void timer_callback()
    {
        RCLCPP_INFO(this->get_logger(), "Time is ticking: %d", counter);
        counter++;
    }

public:
    CustomNode() : Node("custom_node")
    {
        RCLCPP_INFO(this->get_logger(), "CustomNode has been created.");

        timer = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&CustomNode::timer_callback, this));
    }

};

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<CustomNode>();
    RCLCPP_INFO(node->get_logger(), "Hello from CustomNode! This is a custom ROS 2 node in C++.");
    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}

File: src/hello_world_py_pkg/hello_world_py_pkg/custom_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node


class MyCustomNode(Node):
    def __init__(self):
        super().__init__('my_custom_node')
        self.counter_ = 0
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.get_logger().info('My custom node has been started!')

    def timer_callback(self):
        self.get_logger().info(f'Time is ticking: {self.counter_}')
        self.counter_ += 1

def main(args=None):
    rclpy.init(args=args)

    node = MyCustomNode()
    rclpy.spin(node) # Keep the node alive until shutdown is requested - also activates callbacks
    
    rclpy.shutdown()

if __name__ == '__main__':
    main()